### PR TITLE
Send installer output to stderr

### PR DESCRIPTION
**Pull request by yeetfield**
Addresses #45, #16, perhaps #14 

Tested with `target/debug/ozy run [app] > ~/ozy_stdout` with a conda app and observed stderr still being populated and `~/ozy_stdout` remaining empty

### CODE REVIEW

File: rozy/src/utils.rs [hunk 1]
```diff
@@ -9,3 +10,19 @@ pub fn download_to(dest_path: &std::path::PathBuf, url: &str) -> Result<()> {
...
+pub fn run_with_stderr_to_stdout(
+    mut command: std::process::Command,
+) -> Result<std::process::ExitStatus> {
+    let mut spawned = command.spawn().unwrap();
+
+    let stdout = spawned.stdout.as_mut().unwrap();
+    let stdout_reader = BufReader::new(stdout);
+    let stdout_lines = stdout_reader.lines();
+
+    for line in stdout_lines {
```
Comment (mattgodbolt) [inline rozy/src/utils.rs]: If I'm reading this correctly: this doens't pipe stdout to stderr, it buffers stdout and then reads it one line at a time and re-outputs.

Is there no way to control the child process's file descriptors? In regular C style code or python equivalent, you'd redirect the child's stdout to your own stderr, and then the OS just does the right thing.

File: rozy/src/installers/shell.rs [hunk 2]
```diff
@@ -59,7 +60,7 @@ impl Installer for Shell {
         download_to(&file, &self.url)?;
 
         let mut command = std::process::Command::new("/bin/bash");
-
+        command.stdout(std::process::Stdio::piped());
```
Comment (mattgodbolt) [inline rozy/src/installers/shell.rs]: e.g. can this not be something like command.stdout(std::process::stderr) or something (not near a proper computer to look that up)

**Comment by mattgodbolt** [comment]
google found https://stackoverflow.com/questions/42101920/how-to-redirect-child-process-output-to-stderr

File: rozy/src/utils.rs (hunk 1 above)
Comment (yeetfield) [reply rozy/src/utils.rs]: Yeah, I'd copied this code from here: https://stackoverflow.com/a/50444713. Specifying file descriptors is certainly less circuitous and requires no buffering.

**Comment by yeetfield** [comment]
Oh, `from_raw_fd` requires `unsafe`. It seems that `BufReader` is also recommended in the cookbook: https://rust-lang-nursery.github.io/rust-cookbook/os/external.html#continuously-process-child-process-outputs

In practice this does lead to the data being printed to the terminal as the underlying installer produces it (e.g. it isn't the case that the screen is blank for 30 seconds and then all the installer output is dumped in a torrent of data) which is somewhat more encouraging from a memory perspective and doesn't seem to imply that we're accumulating an arbitrary amount of installer output in memory. Instead, we're subject to the size of `BufReader`'s capacity, which is 8KB by default.

As to the question of whether a `BufReader` is required, the `ChildStdout` instance only implements `Read` and does not offer any ergonomic way of iterating over its contents. It appears that `BufReader` exists to facilitate iterating over things that are `Read`

**Comment by mattgodbolt** [comment]
cc @apmorton - this is annoying as it's specifically something unix processes are designed to "just do" :)  Wasting time on buffers reading and writing as a poor FD proxy is a shame, but is obviously better than no output.

This approach currently does line buffering though, am I right? So a progress bar type thing won't work (or will end up at the end as a single line).

**Comment by apmorton** [comment]
Strong preference for "just" connecting the right fds and letting the OS so its thing.

https://docs.rs/os_pipe/latest/os_pipe/fn.dup_stderr.html

Seems like that little package hides it in an os agnostic way.

**Comment by apmorton** [comment]
Where does stderr from the subprocess go? Presumably it also needs to be piped like stdout.

Can we wrap this up in a util function so every installer doesn't have to reimplement this?

File: rozy/src/installers/conda.rs [hunk 3]
```diff
@@ -28,6 +28,7 @@ pub fn conda_install(
     let conda_cache_dir = tempdir()?;
     let mut command = std::process::Command::new(conda_bin);
     command.env("CONDA_PKGS_DIRS", conda_cache_dir.path());
+    command.stdout(Into::<std::process::Stdio>::into(os_pipe::dup_stderr()?));
```
Comment (apmorton) [inline rozy/src/installers/conda.rs]: Does  `Stdio::from(os_pipe...)` not work? I don't really know rust all too well, just saw that in some docs.

Comment (yeetfield) [reply rozy/src/installers/conda.rs]: Ah, I think that should work. They are related traits: https://doc.rust-lang.org/rust-by-example/conversion/from_into.html

File: rozy/src/installers/conda.rs [hunk 4]
```diff
@@ -26,8 +27,9 @@ pub fn conda_install(
     delete_if_exists(to_dir)?;
 
     let conda_cache_dir = tempdir()?;
-    let mut command = std::process::Command::new(conda_bin);
+    let mut command = Command::new(conda_bin);
```
Comment (mattgodbolt) [inline rozy/src/installers/conda.rs]: I wonder if we should wrap this common pattern of "make a new command with stdout->stderr" ?)

**Review by mattgodbolt — APPROVED** [review]
Looks good to me! thanks for your patience.

and "common" = "happened more than twice" I guess hahah

**Comment by yeetfield** [comment]
> Where does stderr from the subprocess go? Presumably it also needs to be piped like stdout.

When we `.spawn()` commands, the child process inherits stderr from the parent: https://doc.rust-lang.org/src/std/process.rs.html#837

**Comment by yeetfield** [comment]
> Can we wrap this up in a util function so every installer doesn't have to reimplement this?

It's only the shell, conda, and pip installers that require it. I'd prefer to keep the set of generic utilities (i.e. things in`utils.rs`) as small as possible as well

**Comment by mattgodbolt** [comment]
Works for me! thanks

**Comment by apmorton** [comment]
If we want to add support back in for post_install (which we probably do), app.rs is going to need it as well.

It also occurs to me we probably _dont_ want to inherit stdin - nothing our installers run should be able to block for input if the user is running interactively.

I think it's probably best to make a wrapper which explicitly sets behavior for stdin, stdout, and stderr and make that the only way we create Command objects so nobody has to think about it going forward if/when new installers are implemented.

**Comment by yeetfield** [comment]
For some reason, setting `command.stdin(Stdio::null())` breaks Conda's progress bar; rather than the bar erasing itself, the entire terminal is filled. Other than that I've created the utility function in `installers.rs` to better scope it to installer logic.

**Comment by yeetfield** [comment]
Will look into the continued need for `post_install` separately, as discussed in person.

**Pull request merged**